lib/rbi/rbs/method_type_translator.rb

```diff
@@ -0,0 +1,125 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RBI
+  module RBS
+    class MethodTypeTranslator
+      extend T::Sig
+
+      class Error < RBI::Error; end
+
+      class << self
+        extend T::Sig
+
+        #: (Method, ::RBS::MethodType) -> Sig
+        def translate(method, type)
+          translator = new(method)
+          translator.visit(type)
+          translator.result
+        end
+      end
+
+      #: Sig
+      attr_reader :result
+
+      #: (Method) -> void
+      def initialize(method)
+        @method = method
+        @result = T.let(Sig.new, Sig)
+      end
+
+      #: (::RBS::MethodType) -> void
+      def visit(type)
+        type.type_params.each do |param|
+          result.type_params << param.name
+        end
+
+        visit_function_type(type.type)
+
+        block = type.block
+        visit_block_type(block) if block
+      end
+
+      private
+
+      #: (::RBS::Types::Block) -> void
+      def visit_block_type(type)
+        block_param = @method.params.grep(RBI::BlockParam).first
+        raise Error, "No block param found" unless block_param
+
+        block_name = block_param.name
+        block_type = T.cast(translate_type(type.type), RBI::Type::Proc)
+
+        bind = type.self_type
+        block_type.bind(translate_type(bind)) if bind
+        block_type = block_type.nilable unless type.required
+
+        @result.params << SigParam.new(block_name, block_type)
+      end
+
+      #: (::RBS::Types::Function) -> void
+      def visit_function_type(type)
+        index = 0
+
+        type.required_positionals.each do |param|
+          result.params << translate_function_param(param, index)
+          index += 1
+        end
+
+        type.optional_positionals.each do |param|
+          result.params << translate_function_param(param, index)
+          index += 1
+        end
+
+        rest_positional = type.rest_positionals
+        if rest_positional
+          result.params << translate_function_param(rest_positional, index)
+          index += 1
+        end
+
+        type.trailing_positionals.each do |param|
+          result.params << translate_function_param(param, index)
+          index += 1
+        end
+
+        type.required_keywords.each do |name, param|
+          result.params << SigParam.new(name.to_s, translate_type(param.type))
+          index += 1
+        end
+
+        type.optional_keywords.each do |name, param|
+          result.params << SigParam.new(name.to_s, translate_type(param.type))
+          index += 1
+        end
+
+        rest_keyword = type.rest_keywords
+        if rest_keyword
+          result.params << translate_function_param(rest_keyword, index)
+        end
+
+        return_type = translate_type(type.return_type)
+        result.return_type = return_type
```

Comment:
Nit: we don't seem to need this variable.
```suggestion
        result.return_type = translate_type(type.return_type)
```